### PR TITLE
KTOR-3539 Migrate server plugins to multiplatform

### PR DESCRIPTION
**Pull request by rsinukov**

### CODE REVIEW

File: ktor-server/ktor-server-plugins/ktor-server-caching-headers/jvmAndNix/src/io/ktor/server/plugins/cachingheaders/CachingHeaders.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+package io.ktor.server.plugins.cachingheaders
+
+import io.ktor.http.*
+import io.ktor.http.content.*
+import io.ktor.server.application.*
+
+/**
+ * A configuration for the [CachingHeaders] plugin
+ */
+public class CachingHeadersConfiguration {
```
Comment (olme04) [inline ktor-server/ktor-server-plugins/ktor-server-caching-headers/jvmAndNix/src/io/ktor/server/plugins/cachingheaders/CachingHeaders.kt:14]: `CachingHeadersConfiguration` -> `CachingHeadersConfig`

File: ktor-server/ktor-server-plugins/ktor-server-cors/jvmAndNix/test/io/ktor/tests/CORSTest.kt [hunk 2]
```diff
@@ -132,10 +131,27 @@ class CORSTest {
...
+                get() = throw Exception("Not implemented")
+
+            override val port: Int
+                get() = port
+
+            override val host: String
+                get() = host
+
+            override val uri: String
+                get() = throw Exception("Not implemented")
```
Comment (e5l) [inline ktor-server/ktor-server-plugins/ktor-server-cors/jvmAndNix/test/io/ktor/tests/CORSTest.kt:148]: Consider using `TODO("Is not supported in mock implementation")`

File: ktor-server/ktor-server-tests/jvmAndNix/test/io/ktor/tests/server/application/ApplicationEventTest.kt [hunk 3]
```diff
@@ -11,7 +11,7 @@ import kotlin.test.*
 @Suppress("DEPRECATION")
 class ApplicationEventTest {
     @Test
-    fun `Checking the number of calls to ApplicationStopPreparing (use withTestApplication)`() {
+    fun checkingTheNumberOfCallsToApplicationStopPreparingUseWithTestApplication() {
```
Comment (e5l) [inline ktor-server/ktor-server-tests/jvmAndNix/test/io/ktor/tests/server/application/ApplicationEventTest.kt:14]: could we rename these tests?

**Review by e5l — CHANGES_REQUESTED** [review]

**Comment by e5l** [comment]
Consider checking `lift` build

File: ktor-server/ktor-server-plugins/ktor-server-cors/jvmAndNix/test/io/ktor/tests/CORSTest.kt [hunk 4]
```diff
@@ -132,10 +131,27 @@ class CORSTest {
...
+
+            override val method: HttpMethod
+                get() = throw Exception("Not implemented")
+
+            override val remoteHost: String
+                get() = throw Exception("Not implemented")
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-cors/jvmAndNix/test/io/ktor/tests/CORSTest.kt:154]: @sonatype-lift ignore

File: ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/BasicAuth.kt [hunk 5]
```diff
@@ -113,7 +113,7 @@ public fun ApplicationRequest.basicAuthenticationCredentials(charset: Charset? =
             }
 
             val userPass = try {
-                authHeader.blob.decodeBase64Bytes().toString(charset ?: Charsets.ISO_8859_1)
+                String(authHeader.blob.decodeBase64Bytes(), charset = charset ?: Charsets.ISO_8859_1)
             } catch (e: Throwable) {
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/BasicAuth.kt:117]: @sonatype-lift ignore

File: ktor-io/js/src/io/ktor/utils/io/charsets/CharsetJS.kt [hunk 6]
```diff
@@ -8,6 +8,17 @@ public actual abstract class Charset(internal val _name: String) {
     public actual abstract fun newEncoder(): CharsetEncoder
     public actual abstract fun newDecoder(): CharsetDecoder
 
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null) return false
+        if (this::class != other::class) return false
```
[the remark this answers is not in the document]
Comment (rsinukov) [reply ktor-io/js/src/io/ktor/utils/io/charsets/CharsetJS.kt]: wow, auto generated equals is weird. good catch

**Review by e5l — APPROVED** [review]

**Pull request merged**